File: halo_radar_visualize/halo_radar_visualize/halo_radar_merge_scan.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import PointCloud2
from marine_sensor_msgs.msg import RadarSector
import numpy as np
import sensor_msgs_py.point_cloud2 as pc2

class HaloRadarMergeScan(Node):
    def __init__(self):
        super().__init__('halo_radar_merge_scan')


        parameters = {
            'single_shot_pointcloud_topic': 'single_shot_radar_pointcloud',
            'radar_input_topic': '/HaloA/data',
            'merged_pointcloud_topic': 'merged_pointcloud'
        }

        for name, value in parameters.items():
            self.declare_parameter(name, value)

        self.pointcloud_subscription = self.create_subscription(
            PointCloud2,
            self.get_parameter("single_shot_pointcloud_topic").value,
            self.listener_callback,
            10)
        
        self.halo_subscription = self.create_subscription(
            RadarSector,
            self.get_parameter('radar_input_topic').value,
            self.radar_data_callback,
            10)
            
        self.pointcloud_publisher = self.create_publisher(
            PointCloud2,
            self.get_parameter("merged_pointcloud_topic").value,
            10)
        
        self.full_stack_pointcloud = []
        self.previous_angle=0
        self.publish_merged_pointcloud = False

    def radar_data_callback(self, msg):
        angle= msg.angle_start
        if angle>self.previous_angle:
            self.get_logger().info(f"Full Scan. Publish the merged pointcloud")
            self.publish_merged_pointcloud = True
        self.previous_angle=angle

    def listener_callback(self, msg):
        points = np.array(list(pc2.read_points(msg, skip_nans=False)))
        if points.size == 0:
            # self.get_logger().info(f"No points found in the PointCloud2 message.")
            return
        self.full_stack_pointcloud.append(points)
        if self.publish_merged_pointcloud:
            self.merge_pointcloud(msg)
            self.publish_merged_pointcloud = False
        

    def merge_pointcloud(self,msg):
            full_pointcloud = np.concatenate(self.full_stack_pointcloud, axis=0)
            header = msg.header
            full_pointcloud_msg = pc2.create_cloud(header, msg.fields, full_pointcloud)
            self.pointcloud_publisher.publish(full_pointcloud_msg)
            self.full_stack_pointcloud = []  # Clear the data for the next stack

def main(args=None):
    rclpy.init(args=args)
    cropper = HaloRadarMergeScan()
    rclpy.spin(cropper)
    cropper.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: halo_radar_visualize/halo_radar_visualize/halo_radar_merge_scan.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "marine_sensor_msgs/msg/radar_sector.hpp"
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>

class HaloRadarMergeScan : public rclcpp::Node {
public:
    HaloRadarMergeScan() : Node("halo_radar_merge_scan"), previous_angle_(0), publish_merged_pointcloud_(false) {
        this->declare_parameter<std::string>("single_shot_pointcloud_topic", "single_shot_radar_pointcloud");
        this->declare_parameter<std::string>("radar_input_topic", "/HaloA/data");
        this->declare_parameter<std::string>("merged_pointcloud_topic", "merged_pointcloud");

        single_shot_pointcloud_topic_ = this->get_parameter("single_shot_pointcloud_topic").as_string();
        radar_input_topic_ = this->get_parameter("radar_input_topic").as_string();
        merged_pointcloud_topic_ = this->get_parameter("merged_pointcloud_topic").as_string();

        pointcloud_subscription_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
            single_shot_pointcloud_topic_, 10, std::bind(&HaloRadarMergeScan::listener_callback, this, std::placeholders::_1));
        
        halo_subscription_ = this->create_subscription<marine_sensor_msgs::msg::RadarSector>(
            radar_input_topic_, 10, std::bind(&HaloRadarMergeScan::radar_data_callback, this, std::placeholders::_1));
        
        pointcloud_publisher_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(merged_pointcloud_topic_, 10);
    }

private:
    void radar_data_callback(const marine_sensor_msgs::msg::RadarSector::SharedPtr msg) {
        float angle = msg->angle_start;
        if (angle > previous_angle_) {
            RCLCPP_INFO(this->get_logger(), "Full Scan. Publish the merged pointcloud");
            publish_merged_pointcloud_ = true;
        }
        previous_angle_ = angle;
    }

    void listener_callback(const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
        pcl::PointCloud<pcl::PointXYZ> cloud;
        pcl::fromROSMsg(*msg, cloud);
        if (cloud.empty()) {
            // RCLCPP_INFO(this->get_logger(), "No points found in the PointCloud2 message.");
            return;
        }
        full_stack_pointcloud_.push_back(cloud);
        if (publish_merged_pointcloud_) {
            merge_pointcloud(msg);
            publish_merged_pointcloud_ = false;
        }
    }

    void merge_pointcloud(const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
        pcl::PointCloud<pcl::PointXYZ> full_pointcloud;
        for (const auto& cloud : full_stack_pointcloud_) {
            full_pointcloud += cloud;
        }
        sensor_msgs::msg::PointCloud2 full_pointcloud_msg;
        pcl::toROSMsg(full_pointcloud, full_pointcloud_msg);
        full_pointcloud_msg.header = msg->header;
        pointcloud_publisher_->publish(full_pointcloud_msg);
        full_stack_pointcloud_.clear();
    }

    std::string single_shot_pointcloud_topic_;
    std::string radar_input_topic_;
    std::string merged_pointcloud_topic_;
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr pointcloud_subscription_;
    rclcpp::Subscription<marine_sensor_msgs::msg::RadarSector>::SharedPtr halo_subscription_;
    rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr pointcloud_publisher_;
    std::vector<pcl::PointCloud<pcl::PointXYZ>> full_stack_pointcloud_;
    float previous_angle_;
    bool publish_merged_pointcloud_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HaloRadarMergeScan>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
